working/0546-patterns/patterns.md

```diff
@@ -0,0 +1,924 @@
+
+**TODO: Allow this as a matcher too?**
+
+### List binder
+
+```
+listBinder ::= "[" binders "]"
+```
+
+Destructures lists. It is a compile-time error if the value's type does not
+implement `List<T>` for some T. Throws a runtime error if the length of the list
+determined by calling `length` is not equal to the number of subpatterns.
+
+Otherwise, extracts a series of elements from the list using the index operator
+and applies them to the corresponding subpatterns.
+
+**TODO: Allow a `...` element to skip elements and match suffixes or ignore extra
+elements.**
+
+**TODO: Can the rest element be captured in a variable? What does it desugar
+to?**
+
+**TODO: Should we loosen this to match against any Iterables by using
+`elementAt()`?**
+
+**TODO: Can the pattern have a type argument? If so, can that have a type
+pattern?**
+
+### Map binder
+
+```
+mapBinder ::= "{" mapBinderEntries "}"
+
+mapBinderEntries ::= mapBinderEntry ( "," mapBinderEntry )* ","?
+mapBinderEntry ::= expression ":" binder
+```
+
+Destructures maps. It is a compile-timer error if the value's type does not
+implement `Map<K,V>` for some K and V. Otherwise:
+
+1.  For each `mapMatcherEntry`:
+
+    1.  Evaluate `expression` and pass to `containsKey()` on the value.
+
+    1.  If this returns `false`, throw a runtime error.
+
+    1.  Otherwise, pass the expression's value to the value's index operator.
+        Apply the entry's binder to the resulting value.
+
+Note that, unlike with lists, a matched map may have additional entries that
+are not bound by the binder.
+
+If it is a compile-time error if any of the entry key expressions are not
+constant expressions.
+
+**TODO: Can the pattern have type arguments? If so, can they have type
+patterns?**
+
+## Tuples and extractors
+
+My favorite language features add powerful, expressive syntax, but with hooks
+that let users control what that syntax does. A simple example is for-in loops:
+
+```dart
+for (var n in foo()) print(n);
+```
+
+This loop could print the elements in a list or a set. It could iterate over
+a numeric range or generate an infinite series of primes. Because the for-in
+loop is implemented in terms of calling methods on Iterable, a user can
+implement the Iterable interface to provide whatever sequence-like behavior
+they want.
+
+Tuples and extractors are like that for pattern matching. They let users write
+patterns that pull arbitrary data out of objects by invoking getters declared by
+the object's class. This is similar to [`unapply()`][unapply] in Scala and the
+[`componentN()`][component] methods in Kotlin.
+
+[unapply]: https://docs.scala-lang.org/tour/extractor-objects.html
+[component]: https://kotlinlang.org/docs/reference/multi-declarations.html
+
+### Tuple patterns
+
+There are two kinds of these patterns. **Tuple patterns** appear as they do in
+other languages, a parenthesized list of subpatterns, like:
+
+```dart
+var (a, b, c) = object;
```

Comment:
If I understand the proposal correctly, this destructures anything with **at least** three fields rather than **exactly** three fields?  That's a little surprising.